### ISSUE TITLE
Record shard movements over time

### ISSUE THREAD

**Issue by DaveCTurner** [elastic/elasticsearch#60747]
Sometimes it is useful to know how shards have moved around the cluster on the way to their current locations. Today there is no simple way to obtain this information, even with `DEBUG`-level logging. 
We (the @elastic/es-distributed team) discussed this today and we propose introducing a new logger for shard movements in order to expose this information in the logs. This need only look at the cluster state and therefore runs only on the elected master, and should log shard-by-shard changes to the routing table.

I'm marking this as `help wanted` to invite external contributions since this isn't a priority for us right now, but note that it may not be a straightforward task to implement this correctly and efficiently.

**Comment by dakrone** [comment]
@DaveCTurner what about adding this to a history index, similar to how we manage ILM and SLM history? Then it could be available regardless of log turnover and without worrying about master node changes, etc?

**Comment by DaveCTurner** [comment]
Yes, that's also a possibility, with a suitable retention policy. There is something of a chicken-and-egg problem when restarting a cluster: we may want to record shards starting up before the history index itself has started up. But I see that keeping this in an index has some big advantages too.

**Comment by dakrone** [comment]
We could always do both also, with the logger being on by default and a history index being opt-in if desired. Lots of options here!

**Comment by DaveCTurner** [comment]
I recently discovered that Metricbeat's [`shard` metricset](https://www.elastic.co/guide/en/beats/metricbeat/current/metricbeat-metricset-elasticsearch-shard.html) effectively dumps the routing table out, one document per shard, every 10s by default. Resolving this issue would let us streamline that. We can also perhaps lean on Metricbeat here and expose something a bit more useful for that kind of monitoring (e.g. shard movements over the last few minutes only).

**Comment by Leaf-Lin** [comment]
Linking two relevant discussions around the beats collecting shard routing table:
- https://github.com/elastic/kibana/issues/130575
- https://github.com/elastic/beats/issues/31349

**Comment by Kiriakos1998** [comment]
Hello @DaveCTurner, if this issue is still open may I work on it?
